beamer/agent.py

```diff
@@ -31,107 +37,152 @@ def _get_deployment_block(contract_info: dict[str, ContractInfo]) -> BlockNumber
+        if not source_chain.request_manager.functions.allowedLps(
+            self._config.account.address
+        ).call():
             raise RuntimeError("Agent address is not whitelisted on RequestManager")
 
-        claim_request_extension = source_request_manager.functions.claimRequestExtension().call()
+    def _setup_direction(
+        self, direction: tuple[ChainId, ChainId], rollups: dict[ChainId, Rollup], l1: Chain
+    ) -> None:
+        source_chain = rollups[direction[0]]
+        target_chain = rollups[direction[1]]
+
+        self._check_rollup(source_chain, target_chain)
+
+        claim_request_extension = (
+            source_chain.request_manager.functions.claimRequestExtension().call()
+        )
 
-        self.context = Context(
+        context = Context(
             requests=Tracker(),
             claims=Tracker(),
-            source_chain_id=source_chain_id,
-            target_chain_id=target_chain_id,
-            request_manager=source_request_manager,
-            fill_manager=target_fill_manager,
+            source_chain_id=source_chain.id,
+            target_chain_id=target_chain.id,
+            request_manager=source_chain.request_manager,
+            fill_manager=target_chain.fill_manager,
             match_checker=self._config.token_match_checker,
             address=self._config.account.address,
             latest_blocks={},
             config=self._config,
-            web3_l1=w3_l1,
+            web3_l1=l1.w3,
             task_pool=self._task_pool,
             claim_request_extension=claim_request_extension,
             l1_resolutions={},
             l1_invalidations={},
         )
+        direction_identifier = f"{direction[0]}->{direction[1]}"
+        event_processor = EventProcessor(context)
+        self._event_monitors[direction[0]]._on_sync_done.append(event_processor.mark_sync_done)
+        self._event_monitors[direction[0]]._on_new_events.append(event_processor.add_events)
+        if source_chain.id != target_chain.id:
+            self._event_monitors[direction[1]]._on_sync_done.append(event_processor.mark_sync_done)
+            self._event_monitors[direction[1]]._on_new_events.append(event_processor.add_events)
+        self._event_processors[direction_identifier] = event_processor
 
-        self._event_processor = EventProcessor(self.context)
-
+    def _init(self) -> None:
+        # Just add one worker, as this effectively serializes the work
+        # This is necessary as we use the account for all resolutions and
+        # would otherwise run into nonce problems
+        self._task_pool = ThreadPoolExecutor(max_workers=1)
+        self._event_processors: dict[str, EventProcessor] = {}
         self._event_monitors: dict[ChainId, EventMonitor] = {}
+        l1 = self._init_l1_chain()
+        rollups = self._init_rollups()
+        rollup_chain_ids = [chain_id for chain_id in rollups]
```

Comment:
couldnt you just do 

`rollups.keys()` (if you want to have it in a list `list(rollups.keys())`